File: erlang/seminars/lzw/src/lzw.erl
-module(lzw).

-compile(export_all).

-define(Alphabet, "abcdefghijklmnopqrstuvwxyz ").


test() ->
    test("this is a test to see if it works").

test(Text) ->
    io:format("~w~n", [Text]),
    Encoded = encode(Text),
    io:format("~p~n", [Encoded]),
    io:format("length of text: ~w~n", [length(Text)]),
    io:format("length of encoded: ~w~n", [length(Encoded)]),
    case decode(Encoded) of
	Text ->
	    true;
	_ ->
	    false
    end.
    

% Encode a text by gradually building the encodiong table. To start
% with, the table only contains the known alphabet. If a sequence of
% characters are not found in the table then it is added to the table
% given a new code.

encode([]) ->
    [];
encode([Char|Rest]) ->
    Table = table(),
    Word = [Char],
    {found, Code} = encode_word(Word, Table),
    encode(Rest, Word, Code, Table).

encode([], _Sofar, Code, _Table) ->
    [Code];
encode([Char|Rest], Sofar, Code, Table) ->

    %% This is the tricky part, we have read Sofar and now have Char as
    %% the next character. Notice that Sofar is coded in reverse order,
    %% this is ok - all strings in the table are in reverse order.

    Extended = [Char|Sofar],
    %% io:format("looking for ~s in ~p~n", [Extended, Table]),

    case encode_word(Extended, Table) of

	 {found, Ext} ->
	     %% If the extended word is found we continue to see if can
	     %% find an even longer string. We keep the code for the
	     %% extended word.

             %%io:format("found ~s~n", [Extended]),
	     encode(Rest, Extended, Ext, Table);

	 {notfound, Updated} ->
	    %% If the extended string is not found we output the code
	    %% for the word that we had read sofar.  We then lookup
	    %% the code for the charcter and continue with the rest of
	    %% the text.
	    
	    %% io:format("not found ~s~n", [Extended]),
	    %% io:format("output ~w~n", [Sofar]),
	    Word = [Char],
	    {found, Cd} = encode_word(Word, Table),
	    [Code | encode(Rest, Word, Cd, Updated)]
    end.

encode_word(Word, {N, Words}) ->
    %% This is where we do a lookup of the word and if not found adds
    %% it to the table. We increment N so that we know the code for
    %% the next sequence.

    %%io:format("looking for ~w in ~w~n", [Word, Words]),
    case lists:keysearch(Word, 1, Words) of
	{value, {Word, Code}} ->
	    {found, Code};
	false ->
	    %%io:format("adding ~w ~s~n", [N, lists:reverse(Word)]),
	    {notfound, {N+1, [{Word, N}|Words]}}
    end.


% Decoding is slightly trickier, it is almost straight forward but it
% has a catch. 

decode(Codes) ->
    Table = table(),
    decode(Codes, Table).

decode([], _) ->
    [];
decode([Code], {_, Words}) ->
    {value, {Word, Code}} = lists:keysearch(Code, 2, Words),    
    Word;
decode([Code | Codes], {N, Words}) ->
    %% If we have a code we will of course simply do a lookup using
    %% the table. We take for granted that the word si foudn in the 
    %% table,.

    %% io:format("decoding ~w~n", [Code]),
    {value, {Word, _}} = lists:keysearch(Code, 2, Words),

    %% We then also do a lookup for the next code and will almost
    %% certainly also find this in the table.
   
    [Next|_] = Codes,
    NextChar = case lists:keysearch(Next, 2, Words) of
		   {value, {[Char|_], _}} ->
		       %% If the next code is found, we know that the
		       %% encoder has read first Word and then the
		       %% first character of the next code. This would
		       %% then have been added to the table by the
		       %% encoder - we do the same.
                       Char;
		   false ->
		       %% This is the not so obvious part, if the next
		       %% code is not found - we know that the encoder
		       %% has read Word followed by the first
		       %% character in a sequence that is equal to
		       %% Word. So we return the first character in
		       %% Word.

		       io:format("could not find ~w, I bet it is ~w~n", [Next, N]),
		       
		       [Char|_] = Word,
		       Char
	       end,
    %% io:format("adding ~s as code ~w~n", [Word++[Char], N]),
    Word ++ decode(Codes, {N+1, [{Word++[NextChar], N}|Words]}).


    

table() ->   
    N = length(?Alphabet),
    Strings = lists:map(fun(Char) -> [Char] end, ?Alphabet),
    Map = lists:zip(Strings, lists:seq(1,N,1)),
    {N+1, Map}.

	    

File: exercises/lzw/src/lzw.ex
defmodule LZW do

  ## we need a known alphabet
  @alphabet 'abcdefghijklmnopqrstuvwxyz '

  ##  a demo sequence, only use charters in the alphabet

  def test(), do: test('luluulluululluluuuulululluuu')
  
  def test(text) do
    IO.puts("The string of text: #{text}\n")

    IO.puts("\nThe plain string as charlist:")
    IO.inspect(text, charlists: :as_lists)
    IO.puts("\nLength of text: #{length(text)}")

    table = table(@alphabet)
    
    IO.puts("\nEncoding")
    encoded = encode(text, table)

    IO.puts("\nThe encoded string as charlist:")
    IO.inspect(encoded, charlists: :as_lists)
    IO.puts("\nLength of encoded: #{length(encoded)}\n")

    IO.puts("\nDecoding")
    decoded = decode(encoded, table)
	
    case decoded do
      ^text ->
        IO.puts("\nSUCCESS: The decoded message matches the original string!\n")
      _ ->
        IO.puts("\nERROR: The decoded message does NOT match the original string!\n")
    end
  end

  ## Encode the character sequence given a table.

  def encode([], _), do: []
  def encode([char | rest], table) do
    word = [char]
    {:found, code} = encode_word(word, table)
    encode(rest, word, code, table)
  end

  ## Encode a list of characters given a word that you have seen
  ## sofar, its code and the table of codes.
    
  def encode([], _sofar, code, _table), do: [code]
  def encode([char | rest], sofar, code, table) do

    ## extend the word you have seen with the character
    extended = sofar ++ [char]

    ## if you're lucky it's in the table
    case encode_word(extended, table) do
      {:found, ext} ->
	## yes, continue you might be lucky again
        encode(rest, extended, ext, table)

      {:notfound, updated} ->
	## no, settle for the code you have and start over,
	## note that the table has been updated with a code
	## for the extended word
	sofar = [char]
	## this will allways succeed since sofar is a single char word
        {_, cd} = encode_word(sofar, updated)
        [code | encode(rest, sofar, cd, updated)]
    end
  end

  ## Find the code of a word if it is in the table otherwise add it to
  ## the table.
  
  def encode_word(word, table) do
    case lookup_code(table, word) do
      {_, code} ->
	## if found we return the code
        {:found, code}
      nil ->
	## otherwise we update the table
        {:notfound, add(table, word)}
    end
  end

  
  ## Decode a sequence of codes given a code table.

  def decode([], _), do: []
  
  def decode([code], table) do
    ## this is the last code in the sequence
    {word, _code} = lookup_word(table, code)
    IO.puts("out: #{word}")
    word
  end

  def decode([code | codes], table) do
    {word, _} = lookup_word(table, code)
    updated = decode_update(hd(codes), word, table)
    IO.puts("out: #{word}")
    word ++ decode(codes, updated)
  end

  ## Update the table by adding the word we have found exteded with
  ## what should have been the next character read by the encoder.
  
  def decode_update(next, word, table) do
    char = case lookup_word(table, next) do
	     {found, _} ->
	       hd(found)
	     nil ->
               IO.puts("Could not find the code #{next}")
               hd(word)
	   end
    add(table, word++[char])
  end
  

  ## The table is a simple maping from encoded words to codes. We keep
  ## track of the next code to assign to a word that we add.

  def table(alphabet) do
    n = length(alphabet)
    words = Enum.map(alphabet, fn x -> [x] end)
    codes = Enum.to_list(1..n)
    map = List.zip([words, codes])
    next = n + 1
    {next, map}
  end

  def lookup_code({_, words}, word) do
    List.keyfind(words, word, 0)
  end
  def lookup_word({_, words}, code) do
    List.keyfind(words, code, 1)
  end

  def add({n, words}, word) do
    IO.puts("Adding #{word} as code #{n}")
    {n+1, [{word, n}|words]}
  end
  
  

end
